### ISSUE TITLE
[GCP] Metrics are not grouped by dimension

### ISSUE THREAD

**Issue by tommyers-elastic** [elastic/integrations#6568]
We have noticed occurrences of metric documents being sent to ES which have the same timestamp and dimensions (and metadata). As far as we can see, each GCP metric document contains only a single metric.  

There is code in the GCP metrics module in metricbeat which claims to stop this happening. https://github.com/elastic/beats/blob/main/x-pack/metricbeat/module/gcp/metrics/timeseries.go. It appears there is a bug in that logic that is not performing the grouping at all. 

This was tested on the `gcp.loadbalancing` metricset, which uses `gcp.metrics` metricset under the hood.

**Comment by zmoog** [comment]
Side note after some reading about GCP metrics and TSDB.

I noticed the default value for [index.mapping.dimension_fields.limit](https://www.elastic.co/guide/en/elasticsearch/reference/8.6/tsds-index-settings.html#index-mapping-dimension-fields-limit) is `16`. Some integrations [^1] bumped this value to `32`. This is also mentioned in the [developer_tsdb_migration_guidelines.md](https://github.com/elastic/integrations/blob/9495cd71789829009efca5b220f4e5204606d8ef/docs/developer_tsdb_migration_guidelines.md?plain=1#L28-L36).

Some GCP metrics have a lot of labels. For example, the `gcp.loadbalancing_metrics.https.total_latencies.value` has 18 labels:

- 6 `gcp.labels.metrics.*` labels
- 12 `gcp.labels.resource.*` labels

![CleanShot 2023-06-16 at 16 50 07@2x](https://github.com/elastic/integrations/assets/25941/2abd9012-5a91-4616-8389-696a0eca287c)

For some reasons, there is no `gcp.labels.resource.project_id` field in Elastisearch, so I can only find 17 fields there:

![CleanShot 2023-06-16 at 16 57 06@2x](https://github.com/elastic/integrations/assets/25941/e853b841-d9b1-43a9-8a40-f4f789cf0bc7)


[^1]: InfluxDB and CockroachDB

**Comment by zmoog** [comment]
The grouping mechanism works as designed. 

The grouping is handled by the [timeSeriesGrouped()](https://github.com/elastic/beats/blob/d270536618a469aa9cd96295e3ca5d17fbbbe5c9/x-pack/metricbeat/module/gcp/metrics/timeseries.go#L15) function. The [ID()](https://github.com/elastic/beats/blob/d270536618a469aa9cd96295e3ca5d17fbbbe5c9/x-pack/metricbeat/module/gcp/metrics/timeseries.go#L33) function defines the criteria used to group by the values. 

For the `loadbalancing` metrics, the metricset uses [x-pack/metricbeat/module/gcp/timeseries_metadata_collector.go](https://github.com/elastic/beats/blob/d270536618a469aa9cd96295e3ca5d17fbbbe5c9/x-pack/metricbeat/module/gcp/timeseries_metadata_collector.go#L128-L167) as the implementation of the `ID()` function.

By default, the `ID()` function generates ID values like this one:

```json
{
  "metric": {
    "labels": {
      "cache_result": "DISABLED",
      "client_country": "United States",
      "protocol": "HTTP/1.1",
      "proxy_continent": "America",
      "response_code": "200",
      "response_code_class": "200"
    },
    "type": "loadbalancing.googleapis.com/https/request_count" <——— 👀 —————
  },
  "resource": {
    "labels": {
      "backend_name": "giz-2-http-lb-0",
      "backend_scope": "us-central1-a",
      "backend_scope_type": "ZONE",
      "backend_target_name": "giz-2-http-lb",
      "backend_target_type": "BACKEND_SERVICE",
      "backend_type": "INSTANCE_GROUP",
      "forwarding_rule_name": "giz-2-https-lb",
      "matched_url_path_rule": "UNMATCHED",
      "project_id": "elastic-obs-integrations-dev",
      "region": "global",
      "target_proxy_name": "giz-2-https-lb",
      "url_map_name": "giz-2-https-lb"
    },
    "type": "https_lb_rule"
  },
  "timestamp": 1687277100000000000
}
```

Since the ID value contains the metric type `loadbalancing.googleapis.com/https/request_count`, there is no way it would group two different metrics.

As a quick experiment, I commented the [metric.type](https://github.com/elastic/beats/blob/d270536618a469aa9cd96295e3ca5d17fbbbe5c9/x-pack/metricbeat/module/gcp/timeseries_metadata_collector.go#L138-L140) definition in the ID function, generating ID values like this one:

```json
{
  "metric": {
    "labels": {
      "cache_result": "DISABLED",
      "client_country": "United States",
      "protocol": "HTTP/1.1",
      "proxy_continent": "America",
      "response_code": "200",
      "response_code_class": "200"
    }
  },
  "resource": {
    "labels": {
      "backend_name": "giz-2-http-lb-0",
      "backend_scope": "us-central1-a",
      "backend_scope_type": "ZONE",
      "backend_target_name": "giz-2-http-lb",
      "backend_target_type": "BACKEND_SERVICE",
      "backend_type": "INSTANCE_GROUP",
      "forwarding_rule_name": "giz-2-https-lb",
      "matched_url_path_rule": "UNMATCHED",
      "project_id": "elastic-obs-integrations-dev",
      "region": "global",
      "target_proxy_name": "giz-2-https-lb",
      "url_map_name": "giz-2-https-lb"
    },
    "type": "https_lb_rule"
  },
  "timestamp": 1687279560000000000
}
```

With this ID, I expect a group by resource type with single metric types to go together.

Here's the final result in the debugger:

![CleanShot 2023-06-20 at 18 44 32@2x](https://github.com/elastic/integrations/assets/25941/f93865ad-342c-4083-a287-db7ef1696d1d)

Now `https.request.bytes` and `https.request.count` go together.

**Comment by tommyers-elastic** [comment]
thanks for the updates @zmoog. 

"works as designed" seems partially accurate. Do you have any idea why the metric type would have been intentionally added to the ID document?

**Comment by constanca-m** [comment]
> @constanca-m, can consider the fields metric.labels.* and resource.label.* as dimensions?

I believe so. Every field that is under `gcp.dimensions.* ` works for the ID. But I don't know if there is a specific set of fields that identify each data stream.

**Comment by zmoog** [comment]
> I believe so. Every field that is under gcp.dimensions.* works for the ID. But I don't know if there is a specific set of fields that identify each data stream.

Thanks, this is useful: I'll try to align these collectors with loadbalancing.

**Comment by agithomas** [comment]
This problem is relevant to  compute_metrics as well.

In a zone, I have 2 compute instances. The number of documents created is 9 (because there exists no grouping). 

Only two documents have metadata that could relate to the compute instance (gcp.labels.metrics.device_name) . The other documents have metric data which cannot be associated with the compute instance.

Even though this problem is identified as part of TSDB migration, i believe, without having the metadata (gcp.labels.metrics.device_name) in all documents (either by grouping or by copying), the documents does not serve its purpose.

**Comment by tommyers-elastic** [comment]
@agithomas thanks for the heads up on that. does the metadata issue you refer to relate to [this issue](https://github.com/elastic/beats/pull/35090) at all? as a side conversation, what are your thoughts on the linked issue - should we alias the fields to ECS and leave the originals in place?

**Comment by agithomas** [comment]
> does the metadata issue you refer to relate to [this issue](https://github.com/elastic/beats/pull/35090) at all?

The issue I mentioned is not related to it. Sorry that i didn't explain it clearly. Assume, we have a set of metrics, labels and its values such as below.
```
{
  metric1 : value1,
  label_metric1: label_value1,
  metric2: value2,
  label_metric2: label_value2,
  metric3: value,
  label_metric3 : label_value3,
  device : <vm-name>
}
```
what I observed is we have 3 documents
First:
```
{
  metric1 : value1
  metric-label1: label_value1,
  device : <vm-name>
}
```
Second
```
{
  metric2 : value2,
  metric-label2: label_value2
}
```

Third one
```
{
  metric3 :value3,
  metric-label3: label_value3
}
```
When the documents are stored as 3 documents, what i observed is, field `device` is missing from document 2 and document 3. 

Either we must target the metric labels copied into all documents or we must group related documents into one with all labels within it. 


For TSDB enablement, it is important to uniquely separate one resource from another.  Most likely, resource labels and metric labels are going to be super useful here.  So, if we are removing the metric labels, going with the details of the [this issue](https://github.com/elastic/beats/pull/35090), we might be losing some important information.

**Comment by agithomas** [comment]
The alternate option I can consider here is to use `cloud.*` fields as dimensions and get myself unblocked from this issue.

**Comment by zmoog** [comment]
As [noted](https://github.com/elastic/integrations/issues/6568#issuecomment-1604490150) before, `compute`, `redis`, and `cloudsql` metricset have custom metadata collectors.

It is relatively easy to change the `ID()` function to get a different metrics grouping; however, it's not obvious why the dimensions are designed in the `Metadata()` function. For example, setting the instance ID and name is set in the `cloud.instance` vs. `gcp.labels`. is [intentional](https://github.com/elastic/beats/blob/26827e77650be1a68a88bfb4990e86755896fee3/x-pack/metricbeat/module/gcp/metrics/compute/metadata.go#L72-L78).

**Comment by zmoog** [comment]
@agithomas, what fields are we considering as dimensions for the `compute` metricset?

**Comment by zmoog** [comment]
In general, it seems the metadata collectors move some fields out of `resource.labels` into the equivalent ECS field.

In the previous example, `compute` moves `resource.labels.instance_id` into `cloud.instance.id`.

**Comment by agithomas** [comment]
> @agithomas, what fields are we considering as dimensions for the `compute` metricset?

`cloud.account.id`
`cloud.availability_zone`
`cloud.instance.id`
`cloud.provider`
`cloud.region`

If needed, 

fingerprint processor applied on all `gcp.labels`, as new field, can be considered as dimensions

**Comment by agithomas** [comment]
Under gcp.compute metrics, two documents share the same set of labels. Please see the screenshot below. With above mentioned cloud.* fields and fingerprint() processor applied on all gcp.labels, these two documents would overlap if TSDB is enabled, lead to loosing one of these two documents in a timeseries

**Document 1: This document captures uptime information**

<img width="888" alt="image" src="https://github.com/elastic/integrations/assets/101976829/80b1207c-13c4-4f3e-9838-8a6477d90730">

**Document 2: This document captures firewall stats.**

<img width="1126" alt="image" src="https://github.com/elastic/integrations/assets/101976829/17ccecf3-b53a-45f3-9d59-e0e89b44dcf4">

**Comment by constanca-m** [comment]
Is there any update on this issue @zmoog?

**Comment by tommyers-elastic** [comment]
@constanca-m, maurizio is out on PTO - i said i would take over on this. i'm going to write some guidelines for dimensions across the whole package, then when we have something to check against, we can make sure all the grouping logic is working correctly at the beats level.

**Comment by constanca-m** [comment]
@zmoog or @tommyers-elastic, is this still on progress?

**Comment by zmoog** [comment]
@constanca-m, yes, it is. You can expect an update this week.

**Comment by zmoog** [comment]
I am updating this issue after a while.

I want to share how the collection, grouping, and events emitting work for GCP metrics.

## Overview

The GCP metricset processes data in three main phases:

- Collect time series data from Monitoring API
- Group time series data using the Metadata Collector `ID()` function
- Emit events

I will describe the behavior using the `https.backend_request_bytes_count.value` / `https.backend_request.bytes` metric from the load balancing metrics as an example.


## Collection

The GCP metricset collects time series data for each one of the 29 metrics [configured](https://github.com/elastic/beats/blob/f111cbc73ec40b728299cd457e93ae72a8d7a6b9/x-pack/metricbeat/module/gcp/loadbalancing/manifest.yml#L9-L37).

Here's an excerpt from the serialized version of the API response with the `https.backend_request_bytes_count.value` metric data only:

```text
| 21 |   2 |                                                                                                      |
|    |     | ---------------------------------------------------------------------------------------------------  |
|    |     | metric:                                                                                              |
|    |     |     type: loadbalancing.googleapis.com/https/backend_request_bytes_count                             |
|    |     |     labels:                                                                                          |
|    |     |         cache_result: DISABLED                                                                       |
|    |     |         proxy_continent: Europe                                                                      |
|    |     |         response_code: "502"                                                                         |
|    |     |         response_code_class: "500"                                                                   |
|    |     | resource:                                                                                            |
|    |     |     type: https_lb_rule                                                                              |
|    |     |     labels:                                                                                          |
|    |     |         backend_name: INVALID_BACKEND                                                                |
|    |     |         backend_scope: INVALID_BACKEND                                                               |
|    |     |         backend_scope_type: INVALID_BACKEND                                                          |
|    |     |         backend_target_name: aman-2-http-lb                                                          |
|    |     |         backend_target_type: BACKEND_SERVICE                                                         |
|    |     |         backend_type: INVALID_BACKEND                                                                |
|    |     |         forwarding_rule_name: aman-2-https-lb                                                        |
|    |     |         matched_url_path_rule: UNMATCHED                                                             |
|    |     |         project_id: elastic-obs-integrations-dev                                                     |
|    |     |         region: global                                                                               |
|    |     |         target_proxy_name: aman-2-https-lb                                                           |
|    |     |         url_map_name: aman-2-https-lb                                                                |
|    |     | metadata: null                                                                                       |
|    |     | metrickind: 2                                                                                        |
|    |     | valuetype: 2                                                                                         |
|    |     | points:                                                                                              |
|    |     |     - interval:                                                                                      |
|    |     |         endtime:                                                                                     |
|    |     |             seconds: 1695760380                                                                      |
|    |     |             nanos: 0                                                                                 |
|    |     |         starttime:                                                                                   |
|    |     |             seconds: 1695760320                                                                      |
|    |     |             nanos: 1000000                                                                           |
|    |     |       value:                                                                                         |
|    |     |         value:                                                                                       |
|    |     |             int64value: 194                                                                          |
|    |     | unit: ""                                                                                             |
|    |     |                                                                                                      |
|    |     | ------------------------------------                                                                 |
|    |     | metric:                                                                                              |
|    |     |     type: loadbalancing.googleapis.com/https/backend_request_bytes_count                             |
|    |     |     labels:                                                                                          |
|    |     |         cache_result: DISABLED                                                                       |
|    |     |         proxy_continent: Europe                                                                      |
|    |     |         response_code: "502"                                                                         |
|    |     |         response_code_class: "500"                                                                   |
|    |     | resource:                                                                                            |
|    |     |     type: https_lb_rule                                                                              |
|    |     |     labels:                                                                                          |
|    |     |         backend_name: INVALID_BACKEND                                                                |
|    |     |         backend_scope: INVALID_BACKEND                                                               |
|    |     |         backend_scope_type: INVALID_BACKEND                                                          |
|    |     |         backend_target_name: giz-2-http-lb                                                           |
|    |     |         backend_target_type: BACKEND_SERVICE                                                         |
|    |     |         backend_type: INVALID_BACKEND                                                                |
|    |     |         forwarding_rule_name: giz-2-https-lb                                                         |
|    |     |         matched_url_path_rule: UNMATCHED                                                             |
|    |     |         project_id: elastic-obs-integrations-dev                                                     |
|    |     |         region: global                                                                               |
|    |     |         target_proxy_name: giz-2-https-lb                                                            |
|    |     |         url_map_name: giz-2-https-lb                                                                 |
|    |     | metadata: null                                                                                       |
|    |     | metrickind: 2                                                                                        |
|    |     | valuetype: 2                                                                                         |
|    |     | points:                                                                                              |
|    |     |     - interval:                                                                                      |
|    |     |         endtime:                                                                                     |
|    |     |             seconds: 1695760380                                                                      |
|    |     |             nanos: 0                                                                                 |
|    |     |         starttime:                                                                                   |
|    |     |             seconds: 1695760320                                                                      |
|    |     |             nanos: 1000000                                                                           |
|    |     |       value:                                                                                         |
|    |     |         value:                                                                                       |
|    |     |             int64value: 300                                                                          |
|    |     | unit: ""                                                                                             |
|    |     |                                                                                                      |
|    |     | ------------------------------------                                                                 |
|    |     |                                                                                                      |
```

The response contains two time series values for the `https/backend_request_bytes_count` metric.

Note that each time series value has:

- one `metric` field with `type` and `labels`.
- one `resource` field with `type` and `labels`.
- Values are stored in the `points` field.

The labels are metadata. We can consider all metric labels dimensions; see https://cloud.google.com/load-balancing/docs/metrics for details.

In addition to metrics labels, many resource labels should also be considered dimensions.

Please consider the `resource.labels.backend_target_name` field values for the two time:

1. `resource.labels.backend_target_name: aman-2-http-lb`
2. `resource.labels.backend_target_name: giz-2-http-lb`

## Grouping (current implementation)

By default, the Metadata Collector `ID()` functions uses the content of `metric.labels.*` and `resource.labels.*` fields (and a few other) for grouping. Since the two `resource.labels.backend_target_name` fields contain different values, the default `ID()` implementation will not group these two time series values.

We can change the Metadata Collector `ID()` implementation to match our requirements, but we need to check the resource label and identify all the labels that are dimensions. 


## Events

Ultimately, the metricset will create one event for each time series group.


## What's next?

We must check the time series data and identify all the dimensions, including the values of `resource.labels.*` fields.

**Comment by zmoog** [comment]
@constanca-m  @gpop63, and I had a sync over Zoom to share our findings and ideas about the grouping issue with TSDB.

## Summary

Let's use a time series example from GCP as a reference:

```text
metric:                                                                                             
    type: loadbalancing.googleapis.com/https/backend_request_bytes_count                            
    labels:                                                                                         
        cache_result: DISABLED                                                                      
        proxy_continent: Europe                                                                     
        response_code: "502"                                                                        
        response_code_class: "500"                                                                  
resource:                                                                                           
    type: https_lb_rule                                                                             
    labels:                                                                                         
        backend_name: INVALID_BACKEND                                                               
        backend_scope: INVALID_BACKEND                                                              
        backend_scope_type: INVALID_BACKEND                                                         
        backend_target_name: giz-2-http-lb                                                          
        backend_target_type: BACKEND_SERVICE                                                        
        backend_type: INVALID_BACKEND                                                               
        forwarding_rule_name: giz-2-https-lb                                                        
        matched_url_path_rule: UNMATCHED                                                            
        project_id: elastic-obs-integrations-dev                                                    
        region: global                                                                              
        target_proxy_name: giz-2-https-lb                                                           
        url_map_name: giz-2-https-lb                                                                
metadata: null                                                                                      
metrickind: 2                                                                                       
valuetype: 2                                                                                        
points:                                                                                             
    - interval:                                                                                     
        endtime:                                                                                    
            seconds: 1695760380                                                                     
            nanos: 0                                                                                
        starttime:                                                                                  
            seconds: 1695760320                                                                     
            nanos: 1000000                                                                          
      value:                                                                                        
        value:                                                                                      
            int64value: 300                                                                         
unit: ""                                                                                      
```

We consider the following fields as dimensions:

- All fields `metric.labels.*`
- All fields `resource.labels.*`
- A selection of the ECS fields, like the ones at https://github.com/elastic/integrations/issues/6568#issuecomment-1614143760.

## Action Plan

Here are the next steps:

1. Update the GCP metricset to group time series by labels (metric and resource) and a subset of ECS fields.
2. Update the integration to fingerprint labels and the subset of ECS fields. 

We opt for fingerprinting to avoid defining any existing label as a dimension and maintain the list as Google Cloud adds new labels. 

### (1) Update the GCP metricset

@gpop63 has already put together a working implementation in this PR: https://github.com/elastic/beats/pull/36682/files (kudos, Gabriel!)

We'll work on the PR to review and test the changes. We'll build a custom Agent to be able to execute the [TSDB-migration-test-kit](https://github.com/elastic/TSDB-migration-test-kit) suite that [[Constança Manteigas]] built.

### (2) Update the integration to fingerprint labels and ECS fields

@constanca-m will look into the integration to add the fingerprinting and run the TSDB-migration-test-kit with the custom agent.

**Comment by zmoog** [comment]
I am testing @gpop63 PR using load balancing metrics.

The metricset collected 44 time series and it grouped them into 16 group.

Here one group with three time series:

```
|     11 |             3 | - key: https.backend_request.bytes                             |
|        |               |   value: 110                                                   |
|        |               |   labels:                                                      |
|        |               |     metrics:                                                   |
|        |               |         cache_result: DISABLED                                 |
|        |               |         proxy_continent: America                               |
|        |               |         response_code: "502"                                   |
|        |               |         response_code_class: "500"                             |
|        |               |     resource:                                                  |
|        |               |         backend_name: INVALID_BACKEND                          |
|        |               |         backend_scope: INVALID_BACKEND                         |
|        |               |         backend_scope_type: INVALID_BACKEND                    |
|        |               |         backend_target_name: aman-2-http-lb                    |
|        |               |         backend_target_type: BACKEND_SERVICE                   |
|        |               |         backend_type: INVALID_BACKEND                          |
|        |               |         forwarding_rule_name: aman-2-http-lb                   |
|        |               |         matched_url_path_rule: UNMATCHED                       |
|        |               |         region: global                                         |
|        |               |         target_proxy_name: aman-2-http-lb                      |
|        |               |         url_map_name: aman-2-https-lb                          |
|        |               |   ecs:                                                         |
|        |               |     cloud:                                                     |
|        |               |         account:                                               |
|        |               |             id: elastic-obs-integrations-dev                   |
|        |               |             name: elastic-obs-integrations-dev                 |
|        |               |         provider: gcp                                          |
|        |               |   timestamp: 2023-09-27T15:54:00Z                              |
|        |               | - key: https.backend_response.bytes                            |
|        |               |   value: 258                                                   |
|        |               |   labels:                                                      |
|        |               |     metrics:                                                   |
|        |               |         cache_result: DISABLED                                 |
|        |               |         proxy_continent: America                               |
|        |               |         response_code: "502"                                   |
|        |               |         response_code_class: "500"                             |
|        |               |     resource:                                                  |
|        |               |         backend_name: INVALID_BACKEND                          |
|        |               |         backend_scope: INVALID_BACKEND                         |
|        |               |         backend_scope_type: INVALID_BACKEND                    |
|        |               |         backend_target_name: aman-2-http-lb                    |
|        |               |         backend_target_type: BACKEND_SERVICE                   |
|        |               |         backend_type: INVALID_BACKEND                          |
|        |               |         forwarding_rule_name: aman-2-http-lb                   |
|        |               |         matched_url_path_rule: UNMATCHED                       |
|        |               |         region: global                                         |
|        |               |         target_proxy_name: aman-2-http-lb                      |
|        |               |         url_map_name: aman-2-https-lb                          |
|        |               |   ecs:                                                         |
|        |               |     cloud:                                                     |
|        |               |         account:                                               |
|        |               |             id: elastic-obs-integrations-dev                   |
|        |               |             name: elastic-obs-integrations-dev                 |
|        |               |         provider: gcp                                          |
|        |               |   timestamp: 2023-09-27T15:54:00Z                              |
|        |               | - key: https.backend_request.count                             |
|        |               |   value: 1                                                     |
|        |               |   labels:                                                      |
|        |               |     metrics:                                                   |
|        |               |         cache_result: DISABLED                                 |
|        |               |         proxy_continent: America                               |
|        |               |         response_code: "502"                                   |
|        |               |         response_code_class: "500"                             |
|        |               |     resource:                                                  |
|        |               |         backend_name: INVALID_BACKEND                          |
|        |               |         backend_scope: INVALID_BACKEND                         |
|        |               |         backend_scope_type: INVALID_BACKEND                    |
|        |               |         backend_target_name: aman-2-http-lb                    |
|        |               |         backend_target_type: BACKEND_SERVICE                   |
|        |               |         backend_type: INVALID_BACKEND                          |
|        |               |         forwarding_rule_name: aman-2-http-lb                   |
|        |               |         matched_url_path_rule: UNMATCHED                       |
|        |               |         region: global                                         |
|        |               |         target_proxy_name: aman-2-http-lb                      |
|        |               |         url_map_name: aman-2-https-lb                          |
|        |               |   ecs:                                                         |
|        |               |     cloud:                                                     |
|        |               |         account:                                               |
|        |               |             id: elastic-obs-integrations-dev                   |
|        |               |             name: elastic-obs-integrations-dev                 |
|        |               |         provider: gcp                                          |
|        |               |   timestamp: 2023-09-27T15:54:00Z                              |
|        |               |                                                                |
|        |               |                                                                |
```

In the last step, the metricset collapsed these three time series into one event:

```text
|      3 | rootfields:                                                        |
|        |     cloud:                                                         |
|        |         account:                                                   |
|        |             id: elastic-obs-integrations-dev                       |
|        |             name: elastic-obs-integrations-dev                     |
|        |         provider: gcp                                              |
|        | modulefields:                                                      |
|        |     labels:                                                        |
|        |         metrics:                                                   |
|        |             cache_result: DISABLED                                 |
|        |             proxy_continent: America                               |
|        |             response_code: "502"                                   |
|        |             response_code_class: "500"                             |
|        |         resource:                                                  |
|        |             backend_name: INVALID_BACKEND                          |
|        |             backend_scope: INVALID_BACKEND                         |
|        |             backend_scope_type: INVALID_BACKEND                    |
|        |             backend_target_name: aman-2-http-lb                    |
|        |             backend_target_type: BACKEND_SERVICE                   |
|        |             backend_type: INVALID_BACKEND                          |
|        |             forwarding_rule_name: aman-2-http-lb                   |
|        |             matched_url_path_rule: UNMATCHED                       |
|        |             region: global                                         |
|        |             target_proxy_name: aman-2-http-lb                      |
|        |             url_map_name: aman-2-https-lb                          |
|        | metricsetfields:                                                   |
|        |     https:                                                         |
|        |         backend_request:                                           |
|        |             bytes: 110                                             |
|        |             count: 1                                               |
|        |         backend_response:                                          |
|        |             bytes: 258                                             |
|        | index: ""                                                          |
|        | id: ""                                                             |
|        | namespace: ""                                                      |
|        | timestamp: 2023-09-27T15:54:00Z                                    |
|        | error: null                                                        |
|        | host: ""                                                           |
|        | service: ""                                                        |
|        | took: 0s                                                           |
|        | period: 0s                                                         |
|        | disabletimeseries: false                                           |
```

cc @constanca-m @agithomas

**Comment by zmoog** [comment]
After some tests using the GKE metrics, @constanca-m discovered an unexpected behavior of the GCP metricset. 

The time series grouping worked as expected, but the TSDB-migration-test-kit detected data loss nonetheless. @constanca-m boiled down the problem to an uncomplicated case: she found two documents with identical timestamps and dimensions that were the cause of data loss when TSDB was enabled.

The clue to understanding what was happening was that the two documents had the same `@timestamp` but different `event.ingested`, with values two minutes apart. This difference suggested that the metricset has not gathered the two documents in the same data collection. 

tl;dr — ingest delay is the cause of this behavior.

We learned that GCP metrics individually have an ingest delay; GCP [describes](https://cloud.google.com/monitoring/api/metrics) ingest delay as:

> Ingest Delay: Data points older than this value are guaranteed to be available to be read, excluding data loss due to errors. The delay does not include any time spent waiting until the next sampling period. The delay, if available, appears at the end of the description text in a sentence of the form “After sampling, data is not visible for up to y seconds.”
The GCP docs includes the above text “After sampling, data is not visible for up to y seconds.” for each metric with an ingest delay.

So, during each collection cycle, the metricset collects data on a different time window for each metric.

Here's how it works.

![image](https://github.com/elastic/integrations/assets/25941/f03ada43-1c18-4427-8421-04165ee8b734)

For example, given `t` as a point in time, the following metric:

- `container/memory.limit.bytes` (no ingest delay) collects data in the “t-1m <—> t” window
- `container/memory/request_bytes` (2 min ingest delay) collects data in the “t-3m <—> t-2m” window

If we have one “data point” in the “t-1m <—> t” time window, the metricset will collect `container/memory.limit.bytes` during "collection 1", and then `container/memory/request_bytes` after two collections, 2 minutes later during "collection 3"

If the data point in the “t-1m <—> t” time window has constant timestamp, this would explain the two documents with the same @timestamp and different `event.ingested` values that @constanca.manteigas found yesterday.

To test this hypothesis, I updated the custom agent image (image id `33ab08e90977`) to add a unique `event.collection_id` field for each collection. This is intended for testing only, it may not be the best solution possible.

@constanca-m is testing the GCP metrics integration, adding `event.collection_id` as a dimension to consider the collection, and she hasn't detected data losses so far.

**Comment by zmoog** [comment]
The tests with the custom agent image ID `33ab08e90977` have been successful. After defining `event.collection_id` field as a dimension, the TSDB-migration-test-kit reported no data loss. 

We moved to the next phase to move the PR https://github.com/elastic/beats/pull/36682/ to the finish line.

Since the `event.collection_id` is a non-standard field, I replaced it with `event.created` because it:

- is an ECS field
- has a proper semantics
- works for our purpose

I pushed a new version of the custom Agent with image ID `92eb8c2b4f30`.

@constanca-m, could you replace `event.collection_id` with `event.created` as a dimension and run the TSDB-migration-test-kit with the Agent image ID `92eb8c2b4f30`?

**Comment by zmoog** [comment]
TSDB does not support the `date` field type for dimensions. Reverted to an ID field; this time, we picked the `event.batch_id`. 

Custom agent image `21c80fd7566b` is available for testing; add `event.batch_id` as a dimension.

**Comment by zmoog** [comment]
A new custom agent build version 8.10.4 image `5bbbdae4689a`) is available for testing

In this version, we replaeced `event.batch_id` with `event.metric_names` as a dimension to avoid data loss from metrics collected with different ingest delay.

See https://github.com/elastic/beats/pull/36682/commits/71c0aa041ed153cf69d2e790ed780e10b0565be3 for more details.

**Comment by zmoog** [comment]
A new custom agent build version 8.10.5 (image `5bbbdae4689a`) is available for testing.

This is just a build after a rebase from main with no other changes. The metric names are not hashed yet, but we can handle this is the pipeline until the next build.

**Comment by zmoog** [comment]
A new custom agent build version 8.10.5 (image `985aaeb1a00f` ) is available for testing.

Changes:
- Replace `event.metric_names` with `event.metric_names_hash`

**Comment by zmoog** [comment]
A new custom agent build version 8.10.5 (image `9c9a3e196ba5` ) is available for testing.

Changes:
- Replace `event.metric_names_hash` with `gcp.metric_names_fingerprint` (https://github.com/elastic/beats/pull/36682/commits/02be686549d540a943a9d3af336acba65090a67c)

**Comment by zmoog** [comment]
@gpop63, a new custom agent build version 8.10.5 (image `178af9f385a4` ) is available for testing the latest change we discussed during the [PR](https://github.com/elastic/beats/pull/36682) review.

Changes:
- Collect metric values after they're all available; the mericset use the largest ingest delay available. 
- Drop the `gcp.metric_names_fingerprint` because it's no longer needed.

**Comment by zmoog** [comment]
We resolved the issue with the merge of PR https://github.com/elastic/beats/pull/36682 and the change will ship in 8.12.0.
